Proposals v2. Iteration3

shamil-gadelshin opened the pull request:
### Iteration 3 implemented features and changes

- grace period
- threshold - a minimum percentage of quorum that must vote to pass the proposal
- clear redundant substrate dependencies
- add maximum allowed active proposals 
- runtime upgrade proposal type
- add error enum in a codex
- update finalized_at proposal field
- change ProposalStatus type
- add stake
- add slashes (status, quorum, threshold). [Issue will be fixed: "Fix stake module and allow slashing and unstaking in the same block."](https://github.com/Joystream/substrate-runtime-joystream/issues/161)
- cancelation fee & rejection fee for the proposer
- stake mocks

shamil-gadelshin committed:
Change proposal approval algorithm

- change approval quorum
- introduce approval threshold
- add tests

shamil-gadelshin committed:
Add proposal grace period

- add proposal parameter
- add proposal status ‘PendingExecution’
- change proposal status and execution algorithms
- add tests

shamil-gadelshin committed:
Simplify proposal tests

- move acquiring proposal_id inside the method

shamil-gadelshin committed:
Update proposal codex module

shamil-gadelshin committed:
Remove unnecessary crates

shamil-gadelshin committed:
Add stake dependency

- add joystream stake & substrate balances dependency
- modify mocks
- add stake to the ProposalParameters

shamil-gadelshin committed:
Add max_active_proposals_number const

shamil-gadelshin committed:
Upgrade TextProposal

- add next text field and tests
- introduce Error enum

shamil-gadelshin committed:
Introduce RuntimeUpgrade proposal type

shamil-gadelshin committed:
Make Text and Wasm length configurable in codex storage

shamil-gadelshin committed:
Merge branch 'proposals_v2' into proposals_v2_iteration3

# Conflicts:
#	modules/proposals/engine/src/lib.rs
#	modules/proposals/engine/src/tests/mod.rs
#	modules/proposals/engine/src/types.rs

shamil-gadelshin committed:
Add MaxActiveProposals storage config parameter

- add MaxActiveProposals storage config parameter
- add comments
- refactor: fix clippy suggestions

shamil-gadelshin committed:
Add stake to the proposal on creation

shamil-gadelshin committed:
Change ProposalStatus type

- introduce ProposalDecisionStatus
- introduce ApprovedProposalStatus
- introduce finalize_proposal()
- remove update_proposal_status()
- change tests

shamil-gadelshin committed:
Add stakes to the proposal

shamil-gadelshin committed:
Add remove_stake() function with errors mapping

shamil-gadelshin committed:
Refactor stakes error handling.

shamil-gadelshin committed:
Add stake removing during proposal finalization

- add remove_stake() call
- introduce FinalizationStatus

shamil-gadelshin committed:
Add test for stakes and balances

shamil-gadelshin committed:
Add stake mocks

shamil-gadelshin committed:
Add stake mock test

shamil-gadelshin committed:
Add comments

shamil-gadelshin committed:
Add proposal slashing with tests

shamil-gadelshin committed:
Refactor create_proposal() ensure’s and add new check.

- add ‘approval threshold parameters > 0’
- add ‘slashing threshold parameters > 0’

shamil-gadelshin committed:
Add slashes

shamil-gadelshin committed:
Refactor types and tests

- add proposal status helpers
- move lib and tests to helpers
- add finalize_proposal_failed_using_stake_mocks() test
- add proposal parameters helper

shamil-gadelshin committed:
Change comments

shamil-gadelshin committed:
Add comments regarding stake module issue

https://github.com/Joystream/substrate-runtime-joystream/issues/161

bedeho reviewed (changes requested):
- Is it the case that this follows the approach of propagating errors due to broken invariants? e.g. in

```Rust
        let stake_id = stake_balance
            .map(|stake_amount| {
                T::StakeHandlerProvider::stakes().create_stake(stake_amount, account_id)
            })
            .transpose()?;
```

Basically the approach of absorbing broken invariants into the errors? If so, then I think its OK for now, but we should review and conclude on this issue based on the summary from all we learned from the Parity discussion, and have a clear rule to follow.

As is, it is statically possible to fail past the marked _safe mutation_ in the code path, is it guaranteed that the only failures we get from calls to staking library are due to broken invariants, or are some still possible without this, in which case there is a bug in the extrinsics checks in this module.

Another issue with the error propagation approach here is that, since the errors are not properly typed, there is no explicit conversion of error types, and so understanding the full up to date view possible errors requires chasing down error strings in the staking module.

- None of the proposal creation modules guard the origin, I think we want something like `EnsureMemershipOrigin`, which returns the member id. Then this id can be passed into the proposal engine, so that it can use this as the proposer id it compares to the result of its `EnsureProposerOrigin`. But perhaps I am going to far ahead here.

shamil-gadelshin committed:
Remove approved_at field from the Proposal

shamil-gadelshin committed:
Move finalized_at to the FinalizationData

- move finalized_at from Proposal to the FinalizationStatus
- rename FinalizationStatus to the FinalizationData
- move out all proposal statuses to the separate module
- fix tests

shamil-gadelshin committed:
Move proposal parameters to the separate module

shamil-gadelshin committed:
Add proposal status comments.

shamil-gadelshin committed:
Refactor stakes

- change StakeHandler trait and its default implementation to simple ‘stake module’ proxy
- introduce ProposalStakeManager to combine calls from the StakeHandler into the simple workflow

shamil-gadelshin committed:
Change config() parameters to the Get<> parameters

- rename proposals.body to the ‘description’
- change config() parameters to the Get<> parameters

shamil-gadelshin committed:
Change config() params to Get<> in the codex

- change config() params to Get<>
- rename ‘body’ to the ‘description’ in proposal types and methods

shamil-gadelshin committed:
Apply cargo fmt

bedeho reviewed (approved):
Nice work!
